### ISSUE TITLE
Missing random folder in pyNEST when compiling NEST 3

### ISSUE THREAD

**Issue by babsey** [compneuronmbu/nest-simulator#22]
After compiling and installing NEST 3, I got error in pyNEST that the module 'nest.random' is not found. 

Is there a special configuration for compilation for including random in pyNEST?

I use openSUSE 15.1 and I added the configuration summary.

```
--------------------------------------------------------------------------------
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Linux
Cross Compiling     : FALSE
C compiler          : /usr/bin/cc
C compiler flags    :  -O2 -Wall -fopenmp -fdiagnostics-color=auto
C++ compiler        : /usr/bin/c++
C++ compiler flags  :  -std=c++11 -O2 -Wall -fopenmp -fdiagnostics-color=auto
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6.5: /usr/bin/python3)
       Includes     : /usr/include/python3.6m
       Libraries    : /usr/lib64/libpython3.6m.so

Cython bindings     : Yes (Cython 0.27.3: /usr/bin/cython)
Use threading       : Yes (OpenMP: -fopenmp)
Use GSL             : Yes (GSL 2.4)
    Includes        : /usr/include
    Libraries       : /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so

Use Readline        : No
Use libltdl         : Yes (LTDL 2.4.6)
    Includes        : /usr/include
    Libraries       : /usr/lib64/libltdl.so

Use doxygen         : No
Use MPI             : No
Use MUSIC           : No
Use libneurosim     : No
Use Boost           : No

--------------------------------------------------------------------------------
```

**Comment by hakonsbm** [comment]
@babsey That's strange! Have you tried deleting everything in the build directory and running cmake again?

**Comment by babsey** [comment]
Yes, I deleted build folder and ran cmake again. The problem remains.

**Comment by hakonsbm** [comment]
I can't reproduce your problem when trying to mimic your configuration, and Travis seems to have no problems either. Can you paste the welcome text when importing nest? And maybe provide the full cmake output?

**Comment by babsey** [comment]
Python output when importing nest
```
Python 3.6.5 (default, Mar 31 2018, 19:45:04) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import nest
[INFO] [2019.3.22 17:12:48 /home/spreizer/Projects/nest-simulator/nestkernel/rng_manager.cpp:224 @ Network::create_rngs_] : Creating default RNGs
[INFO] [2019.3.22 17:12:48 /home/spreizer/Projects/nest-simulator/nestkernel/rng_manager.cpp:270 @ Network::create_grng_] : Creating new default global RNG

              -- N E S T --
  Copyright (C) 2004 The NEST Initiative

 Version: nest3/serialize_cleanup@4ff07890
 Built: Mar 22 2019 17:06:06

 This program is provided AS IS and comes with
 NO WARRANTY. See the file LICENSE for details.

 Problems or suggestions?
   Visit http://www.nest-simulator.org

 Type 'nest.help()' to find out more about NEST.

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/nest/py3-nest-3-serialize-cleanup/lib64/python3.6/site-packages/nest/__init__.py", line 29, in <module>
    from .hl_api import *      # noqa
  File "/opt/nest/py3-nest-3-serialize-cleanup/lib64/python3.6/site-packages/nest/hl_api.py", line 27, in <module>
    from .random.hl_api_random import *
ModuleNotFoundError: No module named 'nest.random'
```


Run the command cmake: `cmake -DCMAKE_INSTALL_PREFIX:PATH=/opt/nest/py3-nest-3-serialize-cleanup -Dwith-python=3 /home/spreizer/Projects/nest-simulator`

```
-- The CXX compiler identification is GNU 7.3.1
-- The C compiler identification is GNU 7.3.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for include file mach-o/dyld.h
-- Looking for include file mach-o/dyld.h - not found
-- Looking for include file mach/mach.h
-- Looking for include file mach/mach.h - not found
-- Looking for include file memory.h
-- Looking for include file memory.h - found
-- Looking for include file stdint.h
-- Looking for include file stdint.h - found
-- Looking for include file sys/types.h
-- Looking for include file sys/types.h - found
-- Looking for C++ include istream
-- Looking for C++ include istream - found
-- Looking for C++ include ostream
-- Looking for C++ include ostream - found
-- Looking for C++ include sstream
-- Looking for C++ include sstream - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of long long
-- Check size of long long - done
-- Check size of u_int16_t
-- Check size of u_int16_t - done
-- Check size of uint16_t
-- Check size of uint16_t - done
-- Check size of u_int64_t
-- Check size of u_int64_t - done
-- Check size of uint64_t
-- Check size of uint64_t - done
-- Looking for NAN
-- Looking for NAN - found
-- Looking for isnan
-- Looking for isnan - found
-- Looking for M_E
-- Looking for M_E - found
-- Looking for M_PI
-- Looking for M_PI - found
-- Looking for expm1
-- Looking for expm1 - not found
-- Host triple: x86_64-pc-linux
-- Target triple: x86_64-pc-linux
-- Found LTDL: /usr/lib64/libltdl.so (found version "2.4.6") 
-- Could NOT find Readline (missing: READLINE_LIBRARY READLINE_INCLUDE_DIR) 
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Found GSL: /usr/include (found version "2.4") 
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.5", minimum required is "3") 
-- Found PythonLibs: /usr/lib64/libpython3.6m.so (found suitable exact version "3.6.5") 
-- Found PythonInterp: /usr/bin/python3 (found version "3.6.5") 
-- Found Cython: /usr/bin/cython (found version "0.27.3") 
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Check the abort exitcode.
-- Check the abort exitcode. 134
-- Check the segmentation fault exitcode.
-- Check the segmentation fault exitcode. 139
-- Check whether the compiler ignores cmath makros.
-- Check whether the compiler ignores cmath makros. OFF
-- Check whether the compiler does NOT include <*.h> headers ISO conformant.
-- Check whether the compiler does NOT include <*.h> headers ISO conformant. OFF
-- Check whether the compiler respects symbolic signal names in signal.h.
-- Check whether the compiler respects symbolic signal names in signal.h. OFF
-- Check static template member declaration.
-- Check static template member declaration. OFF
-- Check for STL vector capacity base unity.
-- Check for STL vector capacity base unity. ON
-- Check for STL vector capacity doubling strategy.
-- Check for STL vector capacity doubling strategy. ON
-- Check whether the compiler fails with ICE.
-- Check whether the compiler fails with ICE. OFF
-- Check if ::nan is available from cmath.
-- Check if ::nan is available from cmath. ON
-- Check if ::isnan is available from cmath.
-- Check if ::isnan is available from cmath. ON
-- Found PythonLibs: /usr/lib64/libpython3.6m.so (found version "3.6.5") 

--------------------------------------------------------------------------------
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Linux
Cross Compiling     : FALSE
C compiler          : /usr/bin/cc
C compiler flags    :  -O2 -Wall -fopenmp -fdiagnostics-color=auto
C++ compiler        : /usr/bin/c++
C++ compiler flags  :  -std=c++11 -O2 -Wall -fopenmp -fdiagnostics-color=auto
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6.5: /usr/bin/python3)
       Includes     : /usr/include/python3.6m
       Libraries    : /usr/lib64/libpython3.6m.so

Cython bindings     : Yes (Cython 0.27.3: /usr/bin/cython)
Use threading       : Yes (OpenMP: -fopenmp)
Use GSL             : Yes (GSL 2.4)
    Includes        : /usr/include
    Libraries       : /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so

Use Readline        : No
Use libltdl         : Yes (LTDL 2.4.6)
    Includes        : /usr/include
    Libraries       : /usr/lib64/libltdl.so

Use doxygen         : No
Use MPI             : No
Use MUSIC           : No
Use libneurosim     : No
Use Boost           : No

--------------------------------------------------------------------------------

The NEST executable will be installed to:
  /opt/nest/py3-nest-3-serialize-cleanup/bin/

Documentation and examples will be installed to:
  /opt/nest/py3-nest-3-serialize-cleanup/share/doc/nest/

PyNEST will be installed to:
    /opt/nest/py3-nest-3-serialize-cleanup/lib64/python3.6/site-packages

To set necessary environment variables, add the following line
to your ~/.bashrc :
  source /opt/nest/py3-nest-3-serialize-cleanup/bin/nest_vars.sh

--------------------------------------------------------------------------------

You can now build and install NEST with
  make
  make install
  make installcheck

If you experience problems with the installation or the use of NEST,
please see http://www.nest-simulator.org/frequently_asked_questions
or go to http://www.nest-simulator.org/community to find out how to
join the user mailing list.

-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/nest-build
```

**Comment by stinebuu** [comment]
@babsey Just to make sure, when you go to ỳour PyNEST install directory, so `/opt/nest/py3-nest-3-serialize-cleanup/lib64/python3.6/site-packages/nest/` is there a folder called random?

**Comment by babsey** [comment]
@stinebuu In installation output I can see whether random folder is installed. In nest site-packages I cannot find a folder called random. 

Yesterday, I was figuring out, why the random folder is not installed. Thus, I used my host OS (openSUSE 15.1), Singularity (Ubuntu 18.04) and Docker (Ubuntu 18.04) and all show the same error. But on my cloud server with Ubuntu 18.04, I was able to install with random folder. I have no idea what is the source of the issue.

Here, I inserted codes for Docker as well as for Singularity, you can build it on your computer.
I also tried with python3-nose, python3-matplotlib python3-pandas, python3-scipy but the error remains. In codes, I removed non-required python packages.


**Docker** with Dockerfile `nest3-serialize_cleanup.Dockerfile`
```
FROM ubuntu:18.04

LABEL maintainer="Sebastian Spreizer <spreizer@web.de>"

RUN apt-get update && apt-get install -y \
    build-essential \
    cmake \
    cython3 \
    ipython3 \
    git \
    libgsl0-dev \
    libltdl7-dev \
    libncurses5-dev \
    libreadline6-dev \
    python3-all-dev \
    python3-numpy

WORKDIR /tmp
RUN rm -rf /tmp/nest-build /tmp/nest-simulator && \
  git clone https://github.com/hakonsbm/nest-simulator.git && \
  cd /tmp/nest-simulator && \
  git fetch && \
  git checkout nest3/serialize_cleanup && \
  mkdir /tmp/nest-build

WORKDIR /tmp/nest-build
RUN cmake -DCMAKE_INSTALL_PREFIX:PATH=/opt/nest-simulator/ -Dwith-python=3 /tmp/nest-simulator && \
    make && \
    make install && \
    rm -rf /tmp/nest-build /tmp/nest-simulator
```

To build 
`docker build -f nest3-serialize_cleanup.Dockerfile -t nest3-serialize_cleanup .
`

To run
`docker run -i -t nest3-serialize_cleanup
`



**Singularity** with the recipes file `nest3-serialize_cleanup.def`
```
Bootstrap: docker
From: ubuntu:18.04

%labels
  AUTHOR <spreizer@web.de>

%post
  apt-get update && apt-get install -y \
    build-essential \
    cmake \
    cython3 \
    ipython3 \
    git \
    libgsl0-dev \
    libltdl7-dev \
    libncurses5-dev \
    libreadline6-dev \
    python3-all-dev \
    python3-numpy

  rm -rf /tmp/nest-build /tmp/nest-simulator
  git clone https://github.com/hakonsbm/nest-simulator /tmp/nest-simulator
  cd /tmp/nest-simulator
  git fetch; git checkout nest3/serialize_cleanup

  mkdir /tmp/nest-build; cd /tmp/nest-build
  cmake -DCMAKE_INSTALL_PREFIX:PATH=/opt/nest-simulator -Dwith-python=3 /tmp/nest-simulator
  make; make install

  rm -rf /tmp/nest-build /tmp/nest-simulator
```

To build
`singularity build nest3-serialize_cleanup.simg nest3-serialize_cleanup.def`

To run
 `singularity shell nest3-serialize_cleanup.simg`

**Comment by hakonsbm** [comment]
@babsey Maybe it's a privilege issue? Try `sudo make install`. Or maybe try installing to another directory in your home directory, for example `/home/spreizer/Projects/nest_install`.

**Comment by stinebuu** [comment]
@babsey Thanks for the docker instructions! I followed your example and I can now reproduce the problem. Very strange! I'll have a look to see if I understand. But I have very little experience with dockers, so I might have to ask you for some guidance.

**Comment by stinebuu** [comment]
@babsey I think I might have fixed it! I am at least able to import `nest.random` on our main `nest-3` branch. So if @hakonsbm can merge that branch into his `nest3/serialize_cleanup` branch, then you can test and see if it works for you?

**Comment by babsey** [comment]
I tested `nest-3` in Singularity and it works. @stinebuu A reward point for you for finding the source of the issue. :smile:

**Comment by stinebuu** [comment]
:smile:

**Comment by babsey** [comment]
Issue solved with  7ee3a47
